google/cloud/spanner/internal/transaction_test.cc

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/internal/transaction.h"
+#include "google/cloud/spanner/internal/time.h"
+#include "google/cloud/spanner/timestamp.h"
+#include "google/cloud/spanner/transaction.h"
+#include <gmock/gmock.h>
+#include <chrono>
+#include <ctime>
+#include <future>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+using TransactionSelector = google::spanner::v1::TransactionSelector;
+
+class KeySet {};
+class ResultSet {};
+
+// Fake `google::cloud::spanner::Client` that supports a single `Read()`
+// operation that does nothing but track the expected transaction callbacks.
+class Client {
+ public:
+  // Set the `read_timestamp` we expect to see, and the `txn_id` we want to
+  // use during the upcoming `Read()` calls.
+  void Reset(Timestamp read_timestamp, std::string txn_id) {
+    read_timestamp_ = read_timestamp;
+    txn_id_ = std::move(txn_id);
+    std::unique_lock<std::mutex> lock(mu_);
+    valid_visits_ = 0;
+  }
+
+  // Return the number of valid visitations made to the transaction during a
+  // completed set of `Read()` calls.
+  int ValidVisits() {
+    std::unique_lock<std::mutex> lock(mu_);
+    return valid_visits_;
+  }
+
+  // User-visible read operation.
+  ResultSet Read(Transaction txn, std::string const& table, KeySet const& keys,
+                 std::vector<std::string> const& columns) {
+    std::function<ResultSet(TransactionSelector&)> read =
+        [this, &table, &keys, &columns](TransactionSelector& selector) {
+          return this->Read(selector, table, keys, columns);
+        };
+    return txn->Visit(std::move(read));
```

Comment:
It looks like access to `Transaction::operator->` and `Transaction::Visit` are the reason that Transaction needs to friend Client. One alternative idea would be to have Transaction friend its own *internal* function instead, and then it would not need to friend `Client`. Maybe something like:

```cc
// transaction.h
namespace internal {
template <typename T>
T Visit(Transaction txn, std::function(T(TransactionSelector&)> f);
}  // namespace internal

class Transaction {
 private:
  template <typename T>
  friend T Visit(........);
};

// Then in this file (same as in client.cc, the unit tests, etc.)
...
  return internal::Visit(txn, std::move(read));
...
```

This would mean that Transaction's friends are all "close friends", meaning that its only friends are classes/functions that live in transaction.h (or a related file that it owns). This would also minimize the amount of faking you need to do in unit tests (e.g., no need to fake a Client class).

Thoughts?

Reply:
Done.  Also renamed `internal::Transaction` to `internal::TransactionImpl` to avoid ambiguity within `internal`.